reduce CPU usage

Stebalien opened the pull request:
After this patch, go-peerstream no longer shows up as a CPU hog.

I would have liked to have removed groups entirely but peerstream doesn't understand the concept of peer IDs so we need *some* form of general-purpose grouping/tagging mechanism. However, we can, and probably should, get rid of groups on streams and listeners to reduce complexity.

Stebalien committed:
remove dead code

Stebalien committed:
use read-locks when possible

Stebalien committed:
index connections by group

We frequently query connections by group (i.e., connections to a peer).

I tried removing the concept of groups entirely (indexing by peer ID) but it
turns out that peerstream really doesn't understand the concept of peer
IDs (it's at a lower level).

fixes https://github.com/libp2p/go-libp2p-swarm/issues/29

(this also makes it possible to set a connections groups on add for convenience)

Stebalien committed:
index connections by net connection

This avoids two `O(n)` iterations over all connections on connect. Really, this
may not be an issue as establishing a connection is often quite CPU
intensive (due to the crypto involved) but it shouldn't hurt.

whyrusleeping reviewed (approved):
LGTM, would be nice to see a test though. If you think the changed code is already tested well enough, go ahead and merge

Kubuxu commented on conn.go:
```diff
@@ -263,15 +263,19 @@ func (s *Swarm) setupConn(netConn tpt.Conn, isServer bool) (*Conn, error) {
-			return c, nil
-		}
+	if c, ok := s.connByNet[netConn]; ok {
+		s.connLock.RUnlock()
```
This should be `Unlock` not `RUnlock`.

whyrusleeping replied:
Yeah, i'd like to see a test that will fail with this bug.

Stebalien replied:
(actually, it shouldn't be there at all; we defer the unlock (copy-paste error...)).

Kubuxu commented on swarm.go:
```diff
@@ -201,20 +202,11 @@ func (s *Swarm) Conns() []*Conn {
-			conns = append(conns, c)
-		}
+	for c := range s.connIdx[g] {
+		conns = append(conns, c)
```
I would change it to:
```go
cs := c.connIdx[g]
conns := make([]*Conn, len(cs))
copy(conns, cs)
```

Stebalien replied:
Fixed (it's a map so I still have to iterate but good catch on the allocation size).

Kubuxu replied:
Ahh, I thought it was an array.

whyrusleeping commented:
Whoops. Maybe reviewing code on my phone isn't the best idea

On Fri, Oct 13, 2017, 10:25 AM Jakub Sztandera <notifications@github.com>
wrote:

> *@Kubuxu* commented on this pull request.
> ------------------------------
>
> In swarm.go
> <https://github.com/libp2p/go-peerstream/pull/13#discussion_r144505375>:
>
> >  func (s *Swarm) ConnsWithGroup(g Group) []*Conn {
>  	s.connLock.RLock()
>  	conns := make([]*Conn, 0, 1)
> -	for c := range s.conns {
> -		if c.InGroup(g) {
> -			conns = append(conns, c)
> -		}
> +	for c := range s.connIdx[g] {
> +		conns = append(conns, c)
>
> I would change it to:
>
> cs := c.connIdx[g]conns := make([]*Conn, len(cs))copy(conns, cs)
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/libp2p/go-peerstream/pull/13#pullrequestreview-69171000>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABL4HDOM3ir2h7_KyUSkcpMcbU2ld1knks5sryx7gaJpZM4P3ugv>
> .
>

Stebalien committed:
remove unnecessary unlock

Stebalien committed:
accurately allocate connections slice

Stebalien commented:
It definitely isn't tested enough (and I should have known better because I fixed at least one "this can't have worked" bugs while working on this...).

Stebalien committed:
test adding the same net conn twice

Stebalien committed:
correctly remove connections from the index.

(tests? who needs those.)

Stebalien committed:
some more leak tests

Stebalien committed:
fix g2 test group

Stebalien committed:
correctly initialize groups when passed to AddConn

Stebalien commented:
"never too small to have bugs" (especially when you turn off the type system and use `interface{}`).

Stebalien commented:
@Kubuxu can I get another review of this (with my changes)? I managed to introduce too many nasty bugs to trust myself.

whyrusleeping commented:
I would feel  more confident if we had more than 30% code coverage here: https://ipfs.io/ipfs/QmcAxXjVHqRsLJ5p7sBLyQMp7UPR98qHsSnbX9DgEWacn2#file0

whyrusleeping commented on conn.go:
```diff
@@ -340,9 +343,35 @@ func (s *Swarm) removeStream(stream *Stream, reset bool) error {
+}
+
+// NOTE: must be called under the connIdxLock lock.
+func (s *Swarm) addConnGroup(c *Conn, g Group) {
```
Does this need to be a method on the Swarm anymore? We seem to be accessing the swarm instance through the `Conn` now anyways

Stebalien replied:
Fair point.

Stebalien committed:
refactor tests

Stebalien committed:
better test coverage

still poor but a lot better

Stebalien committed:
move {add,remove}ConnGroup to Conn (and rename)

Stebalien committed:
update CI

* mostly, add code coverage support

Stebalien committed:
fix race in tests

Stebalien committed:
more complete AddConn test.

Stebalien committed:
test dump/string methods.

Basically, just make sure they don't panic or deadlock.

Stebalien commented:
@whyrusleeping I've added quite a few more tests (53.2% coverage). Unfortunately, we still don't have any tests covering listeners.

Kubuxu commented on conn.go:
```diff
@@ -263,15 +285,18 @@ func (s *Swarm) setupConn(netConn tpt.Conn, isServer bool) (*Conn, error) {
-			return c, nil
-		}
+	if c, ok := s.connByNet[netConn]; ok {
+		return c, nil
```
In this case, do we need to close the stream created in line 273?

Stebalien replied:
Ah. No, actually, in this case we're kind of screwed. If we close the stream muxer, we close the connection. If we leave it open, the connection is in a broken state.

Honestly, we should get rid of these checks entirely. The user should not be adding net connections to the swarm multiple times. If they do, that's a bug in *their* code.

@whyrusleeping do you know why these checks existed in the first place?

whyrusleeping replied:
I think these are no longer necessary, it was really more of a safety measure IIRC

Stebalien replied:
So, I've gone ahead and removed this check entirely in https://github.com/libp2p/go-peerstream/pull/13/commits/0d3d388651b5efde19aa31bf5ea028dfb61e9c8a. Sound reasonable?

Kubuxu reviewed (commented):
I have just one questions. Looks OK otherwise.

Stebalien committed:
don't check if the connection has already been added to the swarm.

There's no good way to handle this (without adding a bunch of extra error
checking). Anyways, it's the job of the programmer to not add connections to the
swarm twice.

whyrusleeping commented on swarm.go:
```diff
@@ -201,20 +200,12 @@ func (s *Swarm) Conns() []*Conn {
-		}
+	cs := s.connIdx[g]
+	conns := make([]*Conn, 0, len(cs))
+	for c := range cs {
```
why not:
```
conns := make([]*Conn, len(cs))
copy(conns, cs)
```
?

Kubuxu replied:
it is a map

whyrusleeping replied:
fine. be that way.

Stebalien replied:
Come to lisp, everything's a list!

whyrusleeping reviewed (approved):
one comment, LGTM otherwise
